tests/test_chiral_restraints.py

```diff
@@ -0,0 +1,506 @@
+    for f in frames:
+        vol = pyramidal_volume(*f[restr_idxs])
+        vols_chiral.append(vol)
+
+    vols_orig = np.array(vols_orig)
+    vols_chiral = np.array(vols_chiral)
+
+    # should be within 5% of 50/50 for the original distribution, a bit redundant
+    assert (np.sum(vols_orig > 0) / len(vols_orig) - 0.5) < 0.05
+    assert (np.sum(vols_orig < 0) / len(vols_orig) - 0.5) < 0.05
+    assert (np.sum(vols_chiral > 0) / len(vols_chiral) - 0.95) < 0.05
+
+    ref_dist = [x for x in vols_orig if x < 0]
+    test_dist = [x for x in vols_chiral if x < 0]
+
+    # should be indistinguishable under KS-test
+    ks, pv = scipy.stats.ks_2samp(ref_dist, test_dist)
+    assert ks < 0.05
+    assert pv > 0.10
+
+    # # useful plotting diagnostics, do not remove.
+    # # compare original distributions
+    # plt.hist(vols_orig, bins=np.linspace(-1, 1, 80), alpha=0.5, label="no_chiral_restr", density=True)
+    # plt.hist(vols_chiral, bins=np.linspace(-1, 1, 80), alpha=0.5, label="with_chiral_restr", density=True)
+    # plt.legend()
+    # plt.title("pyramidal chirality")
+    # plt.xlabel("chiral volume")
+    # plt.ylabel("samples")
+    # plt.title("raw")
+    # plt.show()
+
+    # # time series
+    # plt.plot(vols_chiral, label="with_chiral_restraint")
+    # plt.show()
+
+    # # compare filtered distributions
+    # plt.hist([x for x in vols_orig if x < 0], bins=np.linspace(-1, 1, 80), alpha=0.5, label="no_chiral_restr", density=True)
+    # plt.hist([x for x in vols_chiral if x < 0], bins=np.linspace(-1, 1, 80), alpha=0.5, label="with_chiral_restr", density=True)
+    # plt.legend()
+    # plt.title("pyramidal chirality")
+    # plt.xlabel("chiral volume")
+    # plt.ylabel("samples")
+    # plt.title("normalized")
+    # plt.show()
+
+
+def test_chiral_restraints_torsion():
+    mol = Chem.MolFromMolBlock(
+        """
+  Mrv2202 06062222163D
+
+  4  3  0  0  0  0            999 V2000
+   -0.1757   -0.7570    0.3351 H   0  0  0  0  0  0  0  0  0  0  0  0
+    0.0031   -0.0645   -0.3351 O   0  0  0  0  0  0  0  0  0  0  0  0
+    1.3069    0.0645   -0.3351 O   0  0  0  0  0  0  0  0  0  0  0  0
+    1.4857    0.7570    0.3351 H   0  0  0  0  0  0  0  0  0  0  0  0
+  2  3  1  0  0  0  0
+  1  2  1  0  0  0  0
+  3  4  1  0  0  0  0
+M  END
+$$$$""",
+        removeHs=False,
+    )
+
+    ff = Forcefield.load_from_file("smirnoff_1_1_0_sc.py")
+    ff.q_handle.scale = 0.5
+    s_top = topology.BaseTopology(mol, ff)
+    x0 = utils.get_romol_conf(mol)
+
+    # check initial chirality
+    torsion_idxs = np.array([0, 1, 2, 3])
+    vol = torsion_volume(*x0[torsion_idxs])
+
+    assert abs(vol - 0.0) < 0.05
+    system = s_top.setup_end_state(ff)
+    U_fn = system.get_U_fn()
+
+    vols_orig = []
+    frames = simulate_system(U_fn, x0)
+    for f in frames:
+        vols_orig.append(torsion_volume(*f[torsion_idxs]))
+    vols_orig = np.array(vols_orig)
+
+    assert (np.sum(vols_orig > 0) / len(vols_orig) - 0.22) < 0.05
```

Comment:
Comment about 22/78%?

Reply:
Would it make sense to use abs() to combine the two conditions?

Reply:
willfix